In [24]:
#Let's build a simple convnet to classify the MNIST digits.

# Instantiating a small convnet

from keras import layers
from keras import models

model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))



In [25]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_3 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 3, 3, 64)          36928     
Total params: 55,744
Trainable params: 55,744
Non-trainable params: 0
_________________________________________________________________


In [26]:
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

In [27]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_3 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_1 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_4 (Dense)              (None, 64)               

In [28]:
# Training the convnet on MNIST images

from keras.datasets import mnist
from keras.utils import to_categorical
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [29]:
train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5, batch_size=64)

Epoch 1/5
938/938 [==============================] - 25s 26ms/step - loss: 0.1737 - accuracy: 0.9460
Epoch 2/5
938/938 [==============================] - 25s 26ms/step - loss: 0.0448 - accuracy: 0.9860
Epoch 3/5
938/938 [==============================] - 24s 26ms/step - loss: 0.0317 - accuracy: 0.9897
Epoch 4/5
938/938 [==============================] - 24s 26ms/step - loss: 0.0243 - accuracy: 0.9925
Epoch 5/5
938/938 [==============================] - 25s 26ms/step - loss: 0.0193 - accuracy: 0.9944


In [30]:
# Let’s evaluate the model on the test data:

test_loss, test_acc = model.evaluate(test_images, test_labels)



313/313 [==============================] - 1s 4ms/step - loss: 0.0290 - accuracy: 0.9914


In [31]:
test_acc

0.9914000034332275

In [32]:
test_loss

0.028950316831469536

#Let's now try to classify the MNIST digits by using dense layers instead of convnets

In [9]:
from keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [10]:
train_images.shape



(60000, 28, 28)

In [11]:
len(train_labels)


60000

In [12]:

train_labels


array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [13]:
# And here’s the test data:

test_images.shape


(10000, 28, 28)

In [14]:

len(test_labels)


10000

In [15]:

test_labels


array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

In [16]:
from keras import models
from keras import layers
network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))


In [17]:
network.compile(optimizer='rmsprop',loss='categorical_crossentropy',metrics=['accuracy'])


In [18]:
#Preparing the image data

train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

In [19]:
from keras.utils import to_categorical
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)


In [20]:

network.fit(train_images, train_labels, epochs=5, batch_size=128)


Epoch 1/5
469/469 [==============================] - 2s 5ms/step - loss: 0.2546 - accuracy: 0.9270
Epoch 2/5
469/469 [==============================] - 2s 5ms/step - loss: 0.1029 - accuracy: 0.9690
Epoch 3/5
469/469 [==============================] - 2s 5ms/step - loss: 0.0677 - accuracy: 0.9793
Epoch 4/5
469/469 [==============================] - 2s 5ms/step - loss: 0.0484 - accuracy: 0.9851
Epoch 5/5
469/469 [==============================] - 2s 5ms/step - loss: 0.0371 - accuracy: 0.9890


In [21]:
test_loss, test_acc = network.evaluate(test_images, test_labels)


313/313 [==============================] - 1s 2ms/step - loss: 0.0681 - accuracy: 0.9799


In [22]:
print('test_acc:', test_acc)


test_acc: 0.9799000024795532


Whereas the densely connected network had a test accuracy of approx. 97.9%, the basic convnet has a test accuracy of approx. 99.14%: we decreased the error rate by from 6.8% to 2.89%. Not bad!